bin/download_travis_dependencies.sh

```diff
@@ -1,18 +1,34 @@
 #!/usr/bin/env bash
 
-SPARK_BUILD_URL="http://d3kbcqa49mib13.cloudfront.net/spark-${SPARK_VERSION}-bin-hadoop2.7.tgz"
+set -euo pipefail
 
-echo "Downloading Spark if necessary"
-echo "Spark version = $SPARK_VERSION"
-echo "Spark build = $SPARK_BUILD"
-echo "Spark build URL = $SPARK_BUILD_URL"
-mkdir -p $HOME/.cache/spark-versions
-filename="$HOME/.cache/spark-versions/$SPARK_BUILD.tgz"
-if ! [ -f $filename ]; then
-	echo "Downloading file..."
-	echo `which curl`
-	curl "$SPARK_BUILD_URL" > $filename
-	echo "Content of directory:"
-	ls -la $HOME/.cache/spark-versions/*
-	tar xvf $filename --directory $HOME/.cache/spark-versions > /dev/null
-fi
+SPARK_BUILD="spark-${SPARK_VERSION}-bin-hadoop2.7"
+
+function maybe_download_from_apache_mirror {
+    local spark_tarball="${SPARK_BUILD}.tgz"
+    local apache_mirror_cgi="https://www.apache.org/dyn/closer.lua"
+    local spark_rel_path="spark/spark-${SPARK_VERSION}/${spark_tarball}"
+    local spark_url="${apache_mirror_cgi}?path=${spark_rel_path}"
+
+    echo "Downloading Spark if necessary"
+    echo "Spark version = $SPARK_VERSION"
+    echo "Spark build = $SPARK_BUILD"
+    echo "Spark build URL = $spark_url"
+
+    if [[ ! -f "${spark_tarball}" ]]; then
+        python -- << __PY_SCRIPT_EOF__ | xargs curl -O
+import sys, json
+json_str = """$(curl --silent --location "${spark_url}&asjson=1")"""
+pkg_info = json.loads(json_str)
+print("{}/${spark_rel_path}".format(pkg_info["preferred"]))
+__PY_SCRIPT_EOF__
+
+        echo "Content of directory:"
+        ls -la
+        tar xvf "${spark_tarball}" > /dev/null
```

Comment:
might want to retry if/when curl or tar fails
it actually happens quite a bit

Reply:
That sounds great. I added retry and checksum, etc.